File: backend/app/api/v1/endpoints/integrations.py
"""Unified per-user integration status.

Thin seam that returns one row per supported integration with its
current connection state for the requesting user. Frontend's
Integrations panel renders from this list — the declarative registry
in `frontend/lib/integrations.ts` decides order and copy, the backend
decides status + availability.

Why this exists (2026-04-22): the original design stored each
integration's state on the `user` row (`google_refresh_token`,
`notion_enabled`). That works for 2 integrations, breaks by 5. This
endpoint introduces the forward-compatible shape today — callers see
a typed list — so when we later move to a generic
`integration_connection` table, only the query inside this function
changes. See docs/integrations_architecture.md §Status Endpoint.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.db.models import User
from app.db.scoping import get_current_user_id

router = APIRouter()


def _utc_iso(value: datetime | None) -> str | None:
    """Serialize DB UTC datetimes with an explicit offset.

    The user row stores Moodle sync timestamps as naive UTC. Returning a
    bare ISO string makes browsers parse it as local time, so a fresh
    Cairo sync can look 3h old. Stamp UTC before crossing the API.
    """
    if value is None:
        return None
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    else:
        value = value.astimezone(timezone.utc)
    return value.isoformat()


def _current_user(db: Session) -> User:
    uid = get_current_user_id()
    if uid is None:
        raise HTTPException(status_code=401, detail="not authenticated")
    user = db.query(User).filter(User.user_id == uid).first()
    if user is None:
        raise HTTPException(status_code=401, detail="user not found")
    return user


@router.get("/integrations")
def list_integrations(db: Session = Depends(get_db)) -> dict[str, Any]:
    """List every integration the panel might render with per-user state.

    Status values:
      - 'connected'    — credentials on file, integration actively syncing
      - 'disconnected' — available to connect but not currently linked
      - 'coming_soon'  — displayed as a preview; no connect action yet

    Availability decides whether the frontend renders a Connect button
    vs. a dim "Coming soon" tile. The registry on the frontend is the
    source of truth for human-readable name / description / icon; this
    endpoint only speaks to state.
    """
    user = _current_user(db)

    google_calendar_status = (
        "connected" if user.google_refresh_token else "disconnected"
    )

    # Moodle LMS — connected iff a moodle_ics_url is on file. The
    # disconnect_reason field is surfaced separately so the frontend
    # can show "Reconnect needed" copy when the token went stale and
    # the URL was auto-cleared. (Apr 29 2026, alembic 041 wedge.)
    moodle_status = (
        "connected" if user.moodle_ics_url else "disconnected"
    )

    # Notion write-direction sync is shipped but operator-only. All
    # non-operator users see it as "Coming soon" — we're building user-
    # facing Notion (schema-mapping UI, OAuth) in Phase 7+. When that
    # lands, the `available` branch collapses and this returns real
    # connected/disconnected per user.
    notion_status = "coming_soon"

    # ICS file/URL import is the Phase 7+ Priority 1 integration per
    # docs/import_integrations_capability_map.md. No backend state yet.
    ics_status = "coming_soon"

    return {
        "integrations": [
            {
                "id": "google_calendar",
                "status": google_calendar_status,
                "available": True,
                "scopes": ["https://www.googleapis.com/auth/calendar.readonly"],
            },
            {
                "id": "moodle",
                "status": moodle_status,
                "available": True,
                "scopes": [],
                "last_synced_at": (
                    _utc_iso(user.moodle_last_synced_at)
                ),
                "disconnect_reason": user.moodle_disconnect_reason,
                # Moodle Web Services token (alembic 043, 2026-05-01).
                # Optional sub-capability: when set, the submissions
                # auto-detection sync runs every 6h alongside iCal.
                # Boolean (not the token itself) so /v1/integrations
                # never echoes the credential.
                "ws_connected": bool(user.moodle_ws_token),
                "ws_last_synced_at": (
                    _utc_iso(user.moodle_ws_last_synced_at)
                ),
                "ws_disconnect_reason": user.moodle_ws_disconnect_reason,
            },
            {
                "id": "notion",
                "status": notion_status,
                "available": False,
                "scopes": ["pages:read", "databases:read"],
            },
            {
                "id": "ics",
                "status": ics_status,
                "available": False,
                "scopes": [],
            },
        ]
    }

File: frontend/lib/integrations.ts
/**
 * Integrations registry + client.
 *
 * The frontend's source of truth for human-readable integration
 * metadata (name, description, scope disclosure, "coming soon"
 * previews). Per-user connection state lives server-side and is
 * fetched from GET /v1/integrations — the two join at render time
 * in <IntegrationsSection />.
 *
 * Adding a new integration is ~4 edits:
 *   1. Add an entry to INTEGRATIONS below (registry metadata).
 *   2. Add a case to the backend `list_integrations` endpoint.
 *   3. For OAuth-shaped integrations, add connect + callback routes
 *      under frontend/app/api/integrations/<provider>/.
 *   4. For API-key / file-upload shapes, extend IntegrationCard with
 *      the appropriate input UI.
 *
 * See docs/integrations_architecture.md for the full contract.
 */
import { api } from "./api";

export type IntegrationId =
  | "google_calendar"
  | "moodle"
  | "notion"
  | "ics";

export type IntegrationStatus =
  | "connected"
  | "disconnected"
  | "coming_soon";

export type AuthShape = "oauth" | "api_key" | "file" | "url_subscription";

export interface IntegrationDef {
  id: IntegrationId;
  name: string;
  /** Short one-liner under the name. User-facing. */
  description: string;
  /** What the integration enables. Separate from description so copy
   *  stays tight — description sells the "what", capability sells the "why". */
  capabilityLine: string;
  /** Scope strings shown in the disclosure panel. Empty for file-shape
   *  integrations (no scopes; local parse). */
  scopes: string[];
  authShape: AuthShape;
  /** Present when authShape === "oauth" AND the integration is
   *  available. Connect button is `<a href>` for top-level navigation. */
  connectHref?: string;
  /** Preview tile flag — false means "Coming soon". */
  available: boolean;
  /** Optional note shown on the card — "Phase 7 roadmap", etc. */
  comingSoonNote?: string;
  /** Optional research pre-registration hint. Surfaced only for users
   *  who've opted in to research consent. */
  researchNote?: string;
  /** Two-character monogram rendered in a colored badge. Keeps icon
   *  treatment consistent without bringing in brand logos. */
  monogram: string;
  /** Tailwind utility classes for the monogram chip — card-specific
   *  so each integration has a recognizable color identity. */
  monogramClass: string;
}

export const INTEGRATIONS: IntegrationDef[] = [
  {
    id: "google_calendar",
    name: "Google Calendar",
    description:
      "See your external commitments alongside planned LyraOS tasks.",
    capabilityLine:
      "Read-only — your events appear on /today and /calendar as grey background blocks. Past events carry an optional \"Did you attend?\" self-report.",
    scopes: ["https://www.googleapis.com/auth/calendar.readonly"],
    authShape: "oauth",
    connectHref: "/api/integrations/google-calendar/connect",
    available: true,
    researchNote: "Participates in VT-23 (external-source attendance self-report).",
    monogram: "GC",
    monogramClass:
      "bg-signal/15 text-signal border border-signal/30",
  },
  {
    // Moodle LMS — shipped 2026-04-29 (alembic 041) as the LMS-wedge
    // headline integration. Paste a private .ics subscription URL from
    // Moodle's "Export calendar" page; backend syncs every 6h and
    // upserts assignment due dates as LyraOS deadlines flagged with
    // external_source='moodle_ics'. Tested against ASU Engineering
    // Moodle (lms.eng.asu.edu.eg, Moodle 3.7).
    id: "moodle",
    name: "Moodle",
    description:
      "Pull course deadlines straight from your school's Moodle.",
    capabilityLine:
      "Read-only — assignments and quiz due dates appear as LyraOS deadlines. We won't create posts or modify your courses.",
    scopes: [],
    authShape: "url_subscription",
    available: true,
    monogram: "Mo",
    monogramClass:
      "bg-ember/15 text-ember border border-ember/30",
  },
  {
    id: "notion",
    name: "Notion",
    description:
      "Import tasks from a Notion database; outbound sync already runs for operators.",
    capabilityLine:
      "Column-mapping UI lets you pick which properties become LyraOS's title, date, category. Webhooks + bi-directional sync in v2.",
    scopes: ["pages:read", "databases:read"],
    authShape: "oauth",
    available: false,
    comingSoonNote: "Shipping Phase 7 (post-Spring-School).",
    monogram: "N",
    monogramClass:
      "bg-parchment/10 text-parchment border border-parchment/25",
  },
  {
    id: "ics",
    name: "ICS file / URL",
    description:
      "Drop in any .ics file or subscription URL — Apple, Outlook, Fastmail, or Google export.",
    capabilityLine:
      "Local parse, no third-party auth. Universal fallback for calendars LyraOS doesn't natively integrate with yet.",
    scopes: [],
    authShape: "file",
    available: false,
    comingSoonNote: "Shipping first in Phase 7 — zero-auth import pattern.",
    monogram: "iC",
    monogramClass:
      "bg-dust/10 text-dust border border-dust/25",
  },
];

export interface IntegrationState {
  id: IntegrationId;
  status: IntegrationStatus;
  available: boolean;
  scopes: string[];
  /** Moodle-specific: last successful iCal sync timestamp (ISO).
   *  Surfaced in the card to reassure the user sync is alive. */
  last_synced_at?: string | null;
  /** Moodle-specific: set when the URL was auto-cleared due to a
   *  permanent failure (4xx). Frontend uses this to render
   *  "Reconnect needed" copy instead of plain "Not connected". */
  disconnect_reason?: string | null;
  /** Moodle Web Services token state (alembic 043, 2026-05-01).
   *  Optional sub-capability: when true, the submission-detection
   *  sync records Moodle submission evidence without silently completing
   *  deadlines.
   *  Boolean (never the token itself) so the API never echoes
   *  the credential. */
  ws_connected?: boolean;
  ws_last_synced_at?: string | null;
  ws_disconnect_reason?: string | null;
}

export interface IntegrationsResponse {
  integrations: IntegrationState[];
}

export async function getIntegrations(): Promise<IntegrationsResponse> {
  return api<IntegrationsResponse>("/v1/integrations");
}

/** Shared disconnect helper — used by the Integrations card. Keeps the
 *  legacy `disconnectCalendar` in lib/calendar.ts as the underlying
 *  implementation so we don't fork the endpoint call. */
export async function disconnectIntegration(id: IntegrationId): Promise<void> {
  if (id === "google_calendar") {
    // Re-export the existing endpoint call. Same endpoint, same effect.
    await api("/v1/users/me/google-refresh-token", { method: "DELETE" });
    return;
  }
  if (id === "moodle") {
    await api("/v1/integrations/moodle/disconnect", {
      method: "DELETE",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ void_imported: false }),
    });
    return;
  }
  throw new Error(`disconnect not implemented for ${id}`);
}

// ─── Moodle connect-flow API surface ───────────────────────────────────

export interface MoodlePreviewEvent {
  external_id: string;
  title: string;
  due_at_utc: string; // naive UTC ISO
  category_hint: string | null;
}

export interface MoodlePreviewResponse {
  ok: boolean;
  error: string | null;
  count: number;
  sample: MoodlePreviewEvent[];
}

export interface MoodleConnectResponse {
  ok: boolean;
  preview_count: number;
  sync: {
    fetched: number;
    created: number;
    updated: number;
    unchanged: number;
    skipped_voided: number;
    skipped_unparseable: number;
    duplicate_existing?: number;
    error: string | null;
  };
}

export async function previewMoodle(
  ics_url: string
): Promise<MoodlePreviewResponse> {
  return api<MoodlePreviewResponse>("/v1/integrations/moodle/preview", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ics_url }),
  });
}

export async function connectMoodle(
  ics_url: string
): Promise<MoodleConnectResponse> {
  return api<MoodleConnectResponse>("/v1/integrations/moodle/connect", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ics_url }),
  });
}

export async function syncMoodleNow(): Promise<MoodleConnectResponse["sync"]> {
  const r = await api<MoodleConnectResponse["sync"]>(
    "/v1/integrations/moodle/sync-now",
    { method: "POST" }
  );
  return r;
}


// -----------------------------------------------------------------------
// Moodle Web Services token (Phase B 2026-05-01) — auto-detect submission
// status. Sub-capability of the Moodle integration; iCal can stay
// connected without WS, and vice versa. Backend gates the WS sync to
// users who have set the token.
// -----------------------------------------------------------------------

export interface MoodleWSConnectResponse {
  ok: boolean;
}

export interface MoodleWSSyncResponse {
  ok: boolean;
  matched: number;
  marked_complete: number;
  completion_candidates?: number;
  skipped_no_match: number;
  skipped_not_submitted: number;
  backfilled_completed: number;
  backfilled_completion_candidates?: number;
  backfilled_planned: number;
  backfilled_missed: number;
  marked_titles: string[];
  completion_candidate_titles?: string[];
  backfilled_titles: string[];
  error: string | null;
}

export async function connectMoodleWS(
  ws_token: string,
  base_url?: string,
): Promise<MoodleWSConnectResponse> {
  return api<MoodleWSConnectResponse>("/v1/integrations/moodle/ws-connect", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ws_token, ...(base_url ? { base_url } : {}) }),
  });
}

export async function syncMoodleWSNow(): Promise<MoodleWSSyncResponse> {
  return api<MoodleWSSyncResponse>("/v1/integrations/moodle/ws-sync-now", {
    method: "POST",
  });
}

export async function disconnectMoodleWS(): Promise<void> {
  await api("/v1/integrations/moodle/ws-disconnect", { method: "DELETE" });
}
